webapp/lib/view_tags.dart

```diff
@@ -0,0 +1,251 @@
+        onEditStart: (_) => _title.classes.remove('foldable'),
+        onEditEnd: (_) => _title.classes.add('foldable'),
+        onSave: (_) => controller.command(controller.UIAction.updateTagGroup, new controller.TagGroupData(groupName, newGroupName: _title.text)));
+    _editableTitle.parent = _tagsGroupElement;
+
+    _tagsContainer = new DivElement()..classes.add('tags-group__tags');
+    _tagsGroupElement.append(_tagsContainer);
+
+    _addTagButton = new Button(ButtonType.add,
+        hoverText: 'Add new tag', onClick: (_) => controller.command(controller.UIAction.addTag, new controller.TagData(null, groupId: groupName)));
+    _addTagButton.parent = _tagsContainer;
+
+    var separator = new DivElement()..classes.add('tags-group__separator');
+    _tagsGroupElement.append(separator);
+
+    _title.onClick.listen((event) {
+      _tagsContainer.classes.toggle('hidden');
+      _title.classes.toggle('folded');
+    });
+    // Start off folded
+    _tagsContainer.classes.toggle('hidden', true);
+    _title.classes.toggle('folded', true);
+
+    tagViewsById = {};
+  }
+
+  Element get renderElement => _tagsGroupElement;
+
+  void set name(String name) => _title.text = name;
+
+  void addTags(Map<String, TagView> tags) {
+    for (var tag in tags.keys) {
+      _tagsContainer.insertBefore(tags[tag].renderElement, _addTagButton.renderElement);
+      tagViewsById[tag] = tags[tag];
+    }
+
+    _tagsContainer.style.display = 'block'; // briefly override any display settings to make sure we can compute getBoundingClientRect()
+    bool hidden = _tagsContainer.classes.contains('hidden');
+    _tagsContainer.classes.toggle('hidden', false); // briefly override any display settings to make sure we can compute getBoundingClientRect()
+    List<num> widths = _tagsContainer.querySelectorAll('.tag__name').toList().map((e) => e.getBoundingClientRect().width).toList();
+    num containerWidth = _tagsContainer.getBoundingClientRect().width;
+    _tagsContainer.style.display = ''; // clear inline display settings
+    _tagsContainer.classes.toggle('hidden', hidden);
+
+    num maxGridWidth = widths.fold(0, (previousValue, width) => width > previousValue ? width : previousValue);
+    num colSpacing = 10;
+    num minColWidth = maxGridWidth + 2 * colSpacing + 50;
+    num columnWidth = containerWidth / (containerWidth / minColWidth).floor() - colSpacing;
+    _tagsContainer.style.setProperty('grid-template-columns', 'repeat(auto-fill, ${columnWidth}px)');
+  }
+
+  void modifyTags(Map<String, TagView> tags) {
+    for (var tag in tags.keys) {
+      _tagsContainer.insertBefore(tags[tag].renderElement, tagViewsById[tag].renderElement);
+      tagViewsById[tag].renderElement.remove();
+      tagViewsById[tag] = tags[tag];
+    }
+  }
+
+  void removeTags(List<String> ids) {
+    for (var id in ids) {
+      tagViewsById[id].renderElement.remove();
+      tagViewsById.remove(id);
+    }
+  }
+}
+
+enum TagStyle {
```

Comment:
Worth putting a note in the code where this was forked from so that when we clean this up we'll be able to do so consistently

Reply:
good idea, done!